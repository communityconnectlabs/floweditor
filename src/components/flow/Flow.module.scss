@import "variables.module.scss";

.empty_flow {
  margin: 13px;
  width: 400px;

  h1 {
    font-family: "Helvetica Neue", "RobotoThin", sans-serif;
    letter-spacing: 1px;
    padding: 0;
    text-transform: none;
    font-weight: 100;
    font-size: 18px;
    margin: 0;
  }

  div {
    margin-bottom: 20px;
    font-size: 14px;
    line-height: 18px;
  }
}

.contextMenu {
  z-index: 150;
  border: 1px solid  #ccc;
  background-color: hsl(0,0%,100%);
  border-radius: 5px;
  .menuItem {
    font-size: 14px;
    border-radius: 4px;
    background-color: hsl(0,0%,100%);
    outline: none;
    padding: 10px 16px;
    &:hover {
      color: white;
      background-color: #289f9b;
    }
  }
}
