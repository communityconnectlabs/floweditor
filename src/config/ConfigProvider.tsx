/* eslint-disable @typescript-eslint/no-explicit-any */
import { FlowEditorConfig } from 'flowTypes';
import * as React from 'react';

export const fakePropType: any = (): any => null;
fakePropType.isRequired = (): any => null;

interface ConfigProviderProps {
  config: FlowEditorConfig;
  children: any;
}

export interface ConfigProviderContext {
  config: FlowEditorConfig;
}

// ----------------------------------------------------------------------------------------------

export const SINGLE_CHILD_ERROR = 'ConfigProvider expects only one child component.';
export const VALID_CHILD_ERROR =
  'ConfigProvider expects a valid React element: https://reactjs.org/docs/react-api.html#isvalidelement';

export default class ConfigProvider extends React.Component<ConfigProviderProps> {
  public static childContextTypes = {
    config: fakePropType,
    assetService: fakePropType
  };

  constructor(props: ConfigProviderProps) {
    super(props);
    const win = window as any;
    if (win.isMobile && win.isMobile()) {
      this.props.config.mutable = false;
    }

    if (React.Children.count(props.children) > 1) {
      throw new Error(SINGLE_CHILD_ERROR);
    } else if (!React.isValidElement(props.children)) {
      throw new Error(VALID_CHILD_ERROR);
    }
  }

  public getChildContext(): ConfigProviderContext {
    return {
      config: this.props.config
    };
  }

  public render(): JSX.Element {
    return React.Children.only(this.props.children);
  }
}
