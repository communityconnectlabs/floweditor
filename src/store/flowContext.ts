import { FlowDefinition, FlowNode, UINode } from 'flowTypes';
import { combineReducers } from 'redux';
import ActionTypes, {
  UpdateAssetsAction,
  UpdateBaseLanguageAction,
  UpdateContactFieldsAction,
  UpdateDefinitionAction,
  UpdateDependenciesAction,
  UpdateNodesAction
} from 'store/actionTypes';
import Constants from 'store/constants';

// tslint:disable:no-shadowed-variable
export interface RenderNodeMap {
  [uuid: string]: RenderNode;
}

export interface RenderNode {
  ui: UINode;
  node: FlowNode;
  inboundConnections: { [nodeUUID: string]: string };
  ghost?: boolean;
}

export interface FunctionExample {
  template: string;
  output: string;
}

export interface CompletionOption {
  name?: string;
  summary: string;

  // functions
  signature?: string;
  detail?: string;
  examples?: FunctionExample[];
}

export interface ContactFields {
  [snakedFieldName: string]: string;
}

export enum AssetType {
  Channel = 'channel',
  Classifier = 'classifier',
  Contact = 'contact',
  ContactProperty = 'property',
  Currency = 'currency',
  Environment = 'environment',
  Expression = 'expression',
  Field = 'field',
  Flow = 'flow',
  Global = 'global',
  Group = 'group',
  Label = 'label',
  Language = 'language',
  Remove = 'remove',
  Resthook = 'resthook',
  Result = 'result',
  Revision = 'revision',
  Scheme = 'scheme',
  Template = 'template',
  URN = 'urn'
}

export interface Reference {
  nodeUUID: string;
  actionUUID?: string;
}

export interface Asset {
  id: string;
  name: string;
  type: AssetType;

  order?: number;
  references?: Reference[];
  isNew?: boolean;
  content?: any;
}

export const REMOVE_VALUE_ASSET = {
  id: AssetType.Remove,
  name: 'Remove Value',
  type: AssetType.Remove
};

export const DEFAULT_LANGUAGE = {
  id: 'base',
  name: 'Default',
  type: AssetType.Language
};

export interface AssetStore {
  [assetType: string]: Assets;
}

export interface AssetMap {
  [id: string]: Asset;
}

export interface Assets {
  // our local cache of assets
  items: AssetMap;

  type: AssetType;

  // an optional endpoint to search for more
  endpoint?: string;

  // option name for the id when fetching remotely
  id?: string;

  // have our assets already been loaded
  prefetched?: boolean;
}

export interface FlowContext {
  dependencies: FlowDefinition[];
  baseLanguage: Asset;
  contactFields: ContactFields;
  definition: FlowDefinition;
  nodes: { [uuid: string]: RenderNode };
  assetStore: AssetStore;
}

// Initial state
export const initialState: FlowContext = {
  definition: null,
  dependencies: null,
  baseLanguage: null,
  contactFields: {},
  nodes: {},
  assetStore: {}
};

// Action Creators
export const updateDefinition = (definition: FlowDefinition): UpdateDefinitionAction => ({
  type: Constants.UPDATE_DEFINITION,
  payload: {
    definition
  }
});

export const updateNodes = (nodes: RenderNodeMap): UpdateNodesAction => ({
  type: Constants.UPDATE_NODES,
  payload: {
    nodes
  }
});

export const updateDependencies = (dependencies: FlowDefinition[]): UpdateDependenciesAction => ({
  type: Constants.UPDATE_DEPENDENCIES,
  payload: {
    dependencies
  }
});

export const updateBaseLanguage = (baseLanguage: Asset): UpdateBaseLanguageAction => ({
  type: Constants.UPDATE_BASE_LANGUAGE,
  payload: {
    baseLanguage
  }
});

export const updateContactFields = (contactFields: ContactFields): UpdateContactFieldsAction => ({
  type: Constants.UPDATE_CONTACT_FIELDS,
  payload: {
    contactFields
  }
});

export const updateAssets = (assets: AssetStore): UpdateAssetsAction => ({
  type: Constants.UPDATE_ASSET_MAP,
  payload: {
    assets
  }
});

// Reducers
export const definition = (
  state: FlowDefinition = initialState.definition,
  action: ActionTypes
) => {
  switch (action.type) {
    case Constants.UPDATE_DEFINITION:
      return action.payload.definition;
    default:
      return state;
  }
};

export const nodes = (state: {} = initialState.nodes, action: ActionTypes) => {
  switch (action.type) {
    case Constants.UPDATE_NODES:
      return action.payload.nodes;
    default:
      return state;
  }
};

export const dependencies = (
  state: FlowDefinition[] = initialState.dependencies,
  action: ActionTypes
) => {
  switch (action.type) {
    case Constants.UPDATE_DEPENDENCIES:
      return action.payload.dependencies;
    default:
      return state;
  }
};

export const assetStore = (state: AssetStore = initialState.assetStore, action: ActionTypes) => {
  switch (action.type) {
    case Constants.UPDATE_ASSET_MAP:
      return action.payload.assets;
    default:
      return state;
  }
};

export const baseLanguage = (state: Asset = initialState.baseLanguage, action: ActionTypes) => {
  switch (action.type) {
    case Constants.UPDATE_BASE_LANGUAGE:
      return action.payload.baseLanguage;
    default:
      return state;
  }
};

export const contactFields = (
  state: ContactFields = initialState.contactFields,
  action: ActionTypes
) => {
  switch (action.type) {
    case Constants.UPDATE_CONTACT_FIELDS:
      return action.payload.contactFields;
    default:
      return state;
  }
};

// Root reducer
export default combineReducers({
  definition,
  nodes,
  dependencies,
  assetStore,
  baseLanguage,
  contactFields
});
