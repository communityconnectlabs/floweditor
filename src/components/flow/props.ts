import { UpdateLocalizations } from 'components/nodeeditor/NodeEditor';
import { Type } from 'config/interfaces';
import { AnyAction, ContactProperties } from 'flowTypes';
import { Asset, AssetStore, AssetType, RenderNode } from 'store/flowContext';
import { NodeEditorSettings } from 'store/nodeEditor';
import { DispatchWithState, GetState } from 'store/thunks';
import { titleCase } from 'utils';
import { CompletionSchema } from 'utils/completion';

export interface ActionFormProps {
  // action details
  nodeSettings: NodeEditorSettings;
  typeConfig: Type;
  assetStore: AssetStore;
  completionSchema: CompletionSchema;

  addAsset(assetType: string, asset: Asset): void;

  // update handlers
  updateAction(
    action: AnyAction,
    onUpdated?: (dispatch: DispatchWithState, getState: GetState) => void
  ): void;

  // modal notifiers
  onTypeChange(config: Type): void;
  onClose(canceled: boolean): void;
}

export interface RouterFormProps {
  nodeSettings: NodeEditorSettings;
  typeConfig: Type;

  assetStore: AssetStore;

  // update handlers
  updateRouter(renderNode: RenderNode): void;

  // modal notifiers
  onTypeChange(config: Type): void;
  onClose(canceled: boolean): void;
}

export interface LocalizationFormProps {
  language: Asset;
  nodeSettings: NodeEditorSettings;
  updateLocalizations(languageCode: string, localizations: any[]): UpdateLocalizations;
  onClose(canceled: boolean): void;
}

export const NAME_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Name),
  id: ContactProperties.Name,
  type: AssetType.ContactProperty
};

export const CHANNEL_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Channel),
  id: ContactProperties.Channel,
  type: AssetType.ContactProperty
};

export const LANGUAGE_PROPERTY: Asset = {
  name: titleCase(ContactProperties.Language),
  id: ContactProperties.Language,
  type: AssetType.ContactProperty
};
