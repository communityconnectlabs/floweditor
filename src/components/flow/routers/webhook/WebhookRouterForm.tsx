import { react as bindCallbacks } from 'auto-bind';
import Dialog, { ButtonSet, Tab } from 'components/dialog/Dialog';
import { hasErrors } from 'components/flow/actions/helpers';
import { RouterFormProps } from 'components/flow/props';
import HeaderElement, { Header } from 'components/flow/routers/webhook/header/HeaderElement';
import {
  GET_METHOD,
  METHOD_OPTIONS,
  MethodOption,
  Methods,
  nodeToState,
  stateToNode
} from 'components/flow/routers/webhook/helpers';
import { createResultNameInput } from 'components/flow/routers/widgets';
import SelectElement from 'components/form/select/SelectElement';
import TextInputElement from 'components/form/textinput/TextInputElement';
import { DEFAULT_BODY } from 'components/nodeeditor/constants';
import TypeList from 'components/nodeeditor/TypeList';
import * as React from 'react';
import FlipMove from 'react-flip-move';
import { FormEntry, FormState, mergeForm, StringEntry, ValidationFailure } from 'store/nodeEditor';
import {
  Alphanumeric,
  Required,
  shouldRequireIf,
  StartIsNonNumeric,
  validate,
  ValidURL
} from 'store/validators';
import { createUUID } from 'utils';

import styles from './WebhookRouterForm.module.scss';
import { large } from 'utils/reactselect';

export interface HeaderEntry extends FormEntry {
  value: Header;
}

export interface MethodEntry extends FormEntry {
  value: MethodOption;
}

export interface WebhookRouterFormState extends FormState {
  headers: HeaderEntry[];
  method: MethodEntry;
  url: StringEntry;
  postBody: StringEntry;
  resultName: StringEntry;
}

export default class WebhookRouterForm extends React.Component<
  RouterFormProps,
  WebhookRouterFormState
> {
  constructor(props: RouterFormProps) {
    super(props);
    this.state = nodeToState(this.props.nodeSettings);
    bindCallbacks(this, {
      include: [/^handle/]
    });
  }

  private handleUpdate(
    keys: {
      method?: MethodOption;
      url?: string;
      postBody?: string;
      header?: Header;
      removeHeader?: Header;
      validationFailures?: ValidationFailure[];
      resultName?: string;
    },
    submitting = false
  ): boolean {
    const updates: Partial<WebhookRouterFormState> = {};

    let ensureEmptyHeader = false;

    if (keys.hasOwnProperty('method')) {
      updates.method = { value: keys.method };

      if (keys.method.value !== GET_METHOD.value) {
        if (!this.state.postBody.value) {
          updates.postBody = { value: DEFAULT_BODY };
        }
      } else {
        updates.postBody = { value: null };
      }
    }

    if (keys.hasOwnProperty('url')) {
      updates.url = validate('URL', keys.url, [shouldRequireIf(submitting), ValidURL]);
    }

    if (keys.hasOwnProperty('resultName')) {
      updates.resultName = validate('Result Name', keys.resultName, [shouldRequireIf(submitting)]);
    }

    if (keys.hasOwnProperty('postBody')) {
      updates.postBody = { value: keys.postBody };
    }

    if (keys.hasOwnProperty('header')) {
      updates.headers = [{ value: keys.header, validationFailures: keys.validationFailures }];
      ensureEmptyHeader = true;
    }

    let toRemove: any[] = [];
    if (keys.hasOwnProperty('removeHeader')) {
      toRemove = [{ headers: [{ value: keys.removeHeader }] }];
      ensureEmptyHeader = true;
    }

    const updated = mergeForm(this.state, updates, toRemove);

    // update our form
    this.setState(updated, () => {
      // if we updated headers, check if we need a new one
      if (ensureEmptyHeader) {
        let needsHeader = true;
        for (const header of this.state.headers) {
          if (header.value.name.trim() === '') {
            needsHeader = false;
            break;
          }
        }

        if (needsHeader) {
          this.handleCreateHeader();
        }
      }
    });
    return updated.valid;
  }

  private handleUpdateResultName(value: string): void {
    const resultName = validate('Result Name', value, [Required, Alphanumeric, StartIsNonNumeric]);
    this.setState({
      resultName,
      valid: this.state.valid && !hasErrors(resultName)
    });
  }

  private handleMethodUpdate(method: MethodOption): boolean {
    return this.handleUpdate({ method });
  }

  private handleUrlUpdate(url: string, name: string, submitting = false): boolean {
    return this.handleUpdate({ url }, submitting);
  }

  private handleHeaderRemoved(removeHeader: Header): boolean {
    return this.handleUpdate({ removeHeader });
  }

  private handleHeaderUpdated(header: Header, validationFailures: ValidationFailure[]): boolean {
    return this.handleUpdate({ header, validationFailures });
  }

  private handleCreateHeader(): boolean {
    return this.handleUpdate({
      header: {
        uuid: createUUID(),
        name: '',
        value: ''
      }
    });
  }

  private handlePostBodyUpdate(postBody: string): boolean {
    return this.handleUpdate({ postBody });
  }

  private handleSave(): void {
    // validate our url in case they haven't interacted
    const valid = this.handleUpdate(
      { url: this.state.url.value, resultName: this.state.resultName.value },
      true
    );

    if (valid) {
      this.props.updateRouter(stateToNode(this.props.nodeSettings, this.state));
      this.props.onClose(false);
    }
  }

  private getButtons(): ButtonSet {
    return {
      primary: { name: 'Ok', onClick: this.handleSave },
      secondary: { name: 'Cancel', onClick: () => this.props.onClose(true) }
    };
  }

  private renderEdit(): JSX.Element {
    const typeConfig = this.props.typeConfig;

    const headerElements: JSX.Element[] = this.state.headers.map(
      (header: HeaderEntry, index: number, arr: HeaderEntry[]) => {
        return (
          <div key={`header_${header.value.uuid}`}>
            <HeaderElement
              entry={header}
              onRemove={this.handleHeaderRemoved}
              onChange={this.handleHeaderUpdated}
              index={index}
            />
          </div>
        );
      }
    );

    const tabs: Tab[] = [];
    tabs.push({
      name: 'HTTP Headers',
      hasErrors: !!this.state.headers.find((header: HeaderEntry) => hasErrors(header)),
      body: (
        <>
          <p className={styles.info}>
            Add any additional headers below that you would like to send along with your request.
          </p>
          <FlipMove
            easing="ease-out"
            enterAnimation="elevator"
            leaveAnimation="elevator"
            duration={100}
          >
            {headerElements}
          </FlipMove>
        </>
      ),
      checked: this.state.headers.length > 1
    });

    const method = this.state.method.value.value;
    if (method === Methods.POST || method === Methods.PUT) {
      tabs.push({
        name: 'POST Body',
        body: (
          <div key="post_body" className={styles.body_form}>
            <h4>{this.state.method.value.label} Body</h4>
            <p>Modify the body of your {this.state.method.value.label} request.</p>
            <TextInputElement
              __className={styles.req_body}
              name="Body"
              showLabel={false}
              entry={this.state.postBody}
              onChange={this.handlePostBodyUpdate}
              helpText={`Modify the body of the ${this.state.method.value.label} 
                        request that will be sent to your webhook.`}
              onFieldFailures={(persistantFailures: ValidationFailure[]) => {
                const postBody = { ...this.state.postBody, persistantFailures };
                this.setState({
                  postBody,
                  valid: this.state.valid && !hasErrors(postBody)
                });
              }}
              autocomplete={true}
              textarea={true}
            />
          </div>
        ),
        checked: this.state.postBody.value !== DEFAULT_BODY
      });
    }

    return (
      <Dialog
        title={typeConfig.name}
        headerClass={typeConfig.type}
        buttons={this.getButtons()}
        tabs={tabs}
      >
        <TypeList __className="" initialType={typeConfig} onChange={this.props.onTypeChange} />
        <div className={styles.method}>
          <SelectElement
            styles={large as any}
            name="MethodMap"
            entry={this.state.method}
            onChange={this.handleMethodUpdate}
            options={METHOD_OPTIONS}
          />
        </div>
        <div className={styles.url}>
          <TextInputElement
            name="URL"
            placeholder="Enter a URL"
            entry={this.state.url}
            onChange={this.handleUrlUpdate}
            onFieldFailures={(persistantFailures: ValidationFailure[]) => {
              const url = { ...this.state.url, persistantFailures };
              this.setState({
                url,
                valid: this.state.valid && !hasErrors(url)
              });
            }}
            autocomplete={true}
          />
        </div>
        <div className={styles.instructions}>
          <p>If your server responds with JSON, each property will be added to the Flow.</p>
          <pre className={styles.code}>
            {'{ "product": "Solar Charging Kit", "stock level": 32 }'}
          </pre>
          <p>
            This response would add <span className={styles.example}>@webhook.product</span> and{' '}
            <span className={styles.example}>@webhook["stock level"]</span> for use in the flow.
          </p>
        </div>
        {createResultNameInput(this.state.resultName, this.handleUpdateResultName)}
      </Dialog>
    );
  }

  public render(): JSX.Element {
    return this.renderEdit();
  }
}
