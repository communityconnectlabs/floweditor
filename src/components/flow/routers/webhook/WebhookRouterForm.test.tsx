/* eslint-disable @typescript-eslint/no-object-literal-type-assertion */
/* eslint-disable @typescript-eslint/explicit-function-return-type */
import { RouterFormProps } from 'components/flow/props';
import WebhookRouterForm from 'components/flow/routers/webhook/WebhookRouterForm';
import { Types } from 'config/interfaces';
import { RenderNode } from 'store/flowContext';
import { composeComponentTestUtils, mock } from 'testUtils';
import { createWebhookRouterNode, getRouterFormProps } from 'testUtils/assetCreators';
import * as utils from 'utils';
import * as React from 'react';
import { render, fireEvent, fireChangeText } from 'test/utils';

mock(utils, 'createUUID', utils.seededUUIDs());

const webhookForm = getRouterFormProps({
  node: createWebhookRouterNode(),
  ui: { type: Types.call_webhook }
} as RenderNode);

const { setup } = composeComponentTestUtils<RouterFormProps>(WebhookRouterForm, webhookForm);

describe(WebhookRouterForm.name, () => {
  it('should render', () => {
    const { wrapper } = setup(true);
    expect(wrapper).toMatchSnapshot();
  });

  describe('updates', () => {
    it('should save changes', () => {
      const { baseElement, getByText, getAllByTestId, getByTestId } = render(
        <WebhookRouterForm {...webhookForm} />
      );
      expect(baseElement).toMatchSnapshot();

      const okButton = getByText('Ok');

      // our url is required
      fireEvent.click(okButton);
      expect(webhookForm.updateRouter).not.toBeCalled();

      // set our url and name
      const [url, resultName] = getAllByTestId('input');
      fireChangeText(url, 'http://app.rapidpro.io');
      fireChangeText(resultName, 'My Webhook Result');

      // make it a post
      const selects = getAllByTestId('select');
      fireEvent.change(selects[1], {
        target: { value: 'POST' }
      });

      // set a post body
      fireEvent.click(getByText('POST Body'));
      const postBody = getByTestId('input');
      fireChangeText(postBody, 'Updated post body');

      // add http header
      fireEvent.click(getByText('HTTP Headers'));
      const [headerName, headerValue] = getAllByTestId('input');
      fireEvent.change(headerName, 'Content-type');
      fireEvent.change(headerValue, 'application/json');

      fireEvent.click(okButton);
      expect(webhookForm.updateRouter).toBeCalled();
      expect(webhookForm.updateRouter).toMatchCallSnapshot();
    });

    it('should repopulate post body', () => {
      const { instance } = setup(true, {
        $merge: { onClose: jest.fn(), updateRouter: jest.fn() }
      });

      instance.handleMethodUpdate({ value: 'GET' });
      instance.handleUrlUpdate('http://domain.com/');
      expect(instance.state).toMatchSnapshot();

      instance.handleMethodUpdate('POST');
      expect(instance.state).toMatchSnapshot();
    });

    it('should cancel', () => {
      const { instance, props } = setup(true, {
        $merge: { onClose: jest.fn(), updateRouter: jest.fn() }
      });
      instance.getButtons().secondary.onClick();
      instance.handleUrlUpdate('http://domain.com/');
      expect(props.onClose).toHaveBeenCalled();
      expect(props.updateRouter).not.toHaveBeenCalled();
    });

    it('should validate urls', () => {
      webhookForm.updateRouter = jest.fn();
      const { getByText, getAllByTestId } = render(<WebhookRouterForm {...webhookForm} />);

      // set our url and name
      const [url, resultName] = getAllByTestId('input');
      fireChangeText(url, 'bad url');
      fireChangeText(resultName, 'My Webhook Result');

      // we need a valid url
      const okButton = getByText('Ok');
      fireEvent.click(okButton);
      expect(webhookForm.updateRouter).not.toBeCalled();

      // but not if it has an expression
      fireChangeText(url, '@fields.valid_url');
      fireEvent.click(okButton);
      expect(webhookForm.updateRouter).toBeCalled();
    });
  });
});
