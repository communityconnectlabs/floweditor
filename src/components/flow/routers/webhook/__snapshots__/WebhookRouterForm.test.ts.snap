// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebhookRouterForm should render 1`] = `
<Dialog
  buttons={
    Object {
      "primary": Object {
        "name": "Ok",
        "onClick": [Function],
      },
      "secondary": Object {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="call_webhook"
  tabs={
    Array [
      Object {
        "body": <React.Fragment>
          <p
            className="info"
          >
            Add any additional headers below that you would like to send along with your request.
          </p>
          <FlipMovePropConverter
            delay={0}
            disableAllAnimations={false}
            duration={100}
            easing="ease-out"
            enterAnimation="elevator"
            getPosition={[Function]}
            leaveAnimation="elevator"
            maintainContainerHeight={false}
            staggerDelayBy={0}
            staggerDurationBy={0}
            typeName="div"
            verticalAlignment="top"
          >
            <div>
              <HeaderElement
                entry={
                  Object {
                    "value": Object {
                      "name": "Content-Type",
                      "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
                      "value": "application/json",
                    },
                  }
                }
                index={0}
                onChange={[Function]}
                onRemove={[Function]}
              />
            </div>
            <div>
              <HeaderElement
                entry={
                  Object {
                    "value": Object {
                      "name": "",
                      "uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
                      "value": "",
                    },
                  }
                }
                index={1}
                onChange={[Function]}
                onRemove={[Function]}
              />
            </div>
          </FlipMovePropConverter>
        </React.Fragment>,
        "checked": true,
        "hasErrors": false,
        "name": "HTTP Headers",
      },
    ]
  }
  title="Call Webhook"
>
  <TypeList
    __className=""
    initialType={
      Object {
        "aliases": Array [
          "split_by_webhook",
        ],
        "component": [Function],
        "description": "Call a webhook",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": Array [
          "exits",
        ],
        "name": "Call Webhook",
        "type": "call_webhook",
        "visibility": Array [
          "message",
          "voice",
        ],
      }
    }
    onChange={[MockFunction]}
  />
  <div
    className="method"
  >
    <SelectElement
      entry={
        Object {
          "value": Object {
            "label": "GET",
            "value": "GET",
          },
        }
      }
      name="MethodMap"
      onChange={[Function]}
      options={
        Array [
          Object {
            "label": "GET",
            "value": "GET",
          },
          Object {
            "label": "POST",
            "value": "POST",
          },
          Object {
            "label": "PUT",
            "value": "PUT",
          },
        ]
      }
      styles={
        Object {
          "control": [Function],
          "input": [Function],
          "multiValue": [Function],
          "placeholder": [Function],
          "singleValue": [Function],
        }
      }
    />
  </div>
  <div
    className="url"
  >
    <Connect(TextInputElement)
      autocomplete={true}
      entry={
        Object {
          "value": "",
        }
      }
      name="URL"
      onChange={[Function]}
      onFieldFailures={[Function]}
      placeholder="Enter a URL"
    />
  </div>
  <div
    className="instructions"
  >
    <p>
      If your server responds with JSON, each property will be added to the Flow.
    </p>
    <pre
      className="code"
    >
      { "product": "Solar Charging Kit", "stock level": 32 }
    </pre>
    <p>
      This response would add 
      <span
        className="example"
      >
        @webhook.product
      </span>
       and
       
      <span
        className="example"
      >
        @webhook["stock level"]
      </span>
       for use in the flow.
    </p>
  </div>
  <OptionalTextInput
    helpText="By naming the result, you can reference it later using @results.result"
    maxLength={64}
    name="Result Name"
    onChange={[Function]}
    toggleText="Save as.."
    value={
      Object {
        "value": "Result",
      }
    }
  />
</Dialog>
`;

exports[`WebhookRouterForm updates should repopulate post body 1`] = `
Object {
  "headers": Array [
    Object {
      "value": Object {
        "name": "Content-Type",
        "uuid": "a62e8284-49f8-4292-adbc-d5f6607f85c5",
        "value": "application/json",
      },
    },
    Object {
      "value": Object {
        "name": "",
        "uuid": "0c8c9239-9b59-473a-a229-d8a9693be270",
        "value": "",
      },
    },
  ],
  "method": Object {
    "value": Object {
      "value": "GET",
    },
  },
  "postBody": Object {
    "value": null,
  },
  "resultName": Object {
    "value": "Result",
  },
  "url": Object {
    "validationFailures": Array [],
    "value": "http://domain.com/",
  },
  "valid": true,
}
`;

exports[`WebhookRouterForm updates should repopulate post body 2`] = `
Object {
  "headers": Array [
    Object {
      "value": Object {
        "name": "Content-Type",
        "uuid": "a62e8284-49f8-4292-adbc-d5f6607f85c5",
        "value": "application/json",
      },
    },
    Object {
      "value": Object {
        "name": "",
        "uuid": "0c8c9239-9b59-473a-a229-d8a9693be270",
        "value": "",
      },
    },
  ],
  "method": Object {
    "value": "POST",
  },
  "postBody": Object {
    "value": "@(json(object(
  \\"contact\\", object(
    \\"uuid\\", contact.uuid, 
    \\"name\\", contact.name, 
    \\"urn\\", contact.urn
  ),
  \\"flow\\", object(
    \\"uuid\\", run.flow.uuid, 
    \\"name\\", run.flow.name
  ),
  \\"results\\", foreach_value(results, extract_object, \\"value\\", \\"category\\")
)))",
  },
  "resultName": Object {
    "value": "Result",
  },
  "url": Object {
    "validationFailures": Array [],
    "value": "http://domain.com/",
  },
  "valid": true,
}
`;

exports[`WebhookRouterForm updates should save changes: after header removed 1`] = `
Object {
  "headers": Array [
    Object {
      "value": Object {
        "name": "Content-Type",
        "uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
        "value": "application/json",
      },
    },
    Object {
      "value": Object {
        "name": "",
        "uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
        "value": "",
      },
    },
  ],
  "method": Object {
    "value": Object {
      "value": "POST",
    },
  },
  "postBody": Object {
    "value": "Post Body",
  },
  "resultName": Object {
    "value": "Result",
  },
  "url": Object {
    "persistantFailures": Array [],
    "value": "",
  },
  "valid": true,
}
`;

exports[`WebhookRouterForm updates should save changes: intitial state 1`] = `
Object {
  "headers": Array [
    Object {
      "value": Object {
        "name": "Content-Type",
        "uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
        "value": "application/json",
      },
    },
    Object {
      "value": Object {
        "name": "",
        "uuid": "e67ce051-f3e4-4c42-807b-e393df5fa382",
        "value": "",
      },
    },
    Object {
      "validationFailures": undefined,
      "value": Object {
        "name": "Content-type",
        "uuid": "e0bc3776-792f-49d2-b4be-48be2cf8159a",
        "value": "application/json",
      },
    },
  ],
  "method": Object {
    "value": Object {
      "value": "POST",
    },
  },
  "postBody": Object {
    "value": "Post Body",
  },
  "resultName": Object {
    "value": "Result",
  },
  "url": Object {
    "persistantFailures": Array [],
    "value": "",
  },
  "valid": true,
}
`;

exports[`WebhookRouterForm updates should save changes: save all the things 1`] = `
Array [
  Object {
    "inboundConnections": Object {},
    "node": Object {
      "actions": Array [
        Object {
          "body": "Post Body",
          "headers": Object {
            "Content-Type": "application/json",
          },
          "method": "POST",
          "result_name": "Result",
          "type": "call_webhook",
          "url": "http://domain.com/",
          "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
        },
      ],
      "exits": Array [
        Object {
          "destination_uuid": null,
          "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
        },
        Object {
          "destination_uuid": null,
          "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
        },
      ],
      "router": Object {
        "cases": Array [
          Object {
            "arguments": Array [
              "Success",
            ],
            "category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            "type": "has_only_text",
            "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
          },
        ],
        "categories": Array [
          Object {
            "exit_uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
            "name": "Success",
            "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
          },
          Object {
            "exit_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
            "name": "Failure",
            "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
          },
        ],
        "default_category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
        "operand": "@results.result.category",
        "type": "switch",
      },
      "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
    },
    "ui": Object {
      "config": Object {},
      "position": null,
      "type": "split_by_webhook",
    },
  },
]
`;
